### PR TITLE
Handle non-iterator objects in consume_request_iterator

### LINKED ISSUE #8231: Better support invocation-side erroneously passing non-iterable objects to request-streaming methods
[We don't yet have `for request in request_iterator:` wrapped in a `try:`/`except:`](https://github.com/grpc/grpc/blob/7ab8eecdc9c0c2b1ebb2e69d7460bebca774ebd9/src/python/grpcio/grpc/_channel.py#L184) but we certainly should. If the invocation-side application makes a mistake and passes an object that is not an iterable of request values, something (probably logging) should make them aware of that.

### PR DESCRIPTION
**Pull request by thunderboltsid**
Restructure the consume_request_iterator method to handle exceptions
where the object being passed is not an iterator. Fixes #8231 and #8454.

### CODE REVIEW

**Comment by nathanielmanistaatgoogle** [comment]
Content looks good.

I'm surprising myself by leaning against asking for test coverage of this. What do you think? Is it worth writing a special object that isn't an iterator and confirming that the service-side code observes cancellation and the invocation-side code observes UNKNOWN status?

**Comment by nathanielmanistaatgoogle** [comment]
Please create a `_invocation_defects_test.py` and add test coverage for this enhancement there.

**Comment by nathanielmanistaatgoogle** [comment]
@ncteisen: you should know that this is going on as I think it will conflict with what you're doing to address [issue 8814](https://github.com/grpc/grpc/issues/8814).

File: src/python/grpcio/grpc/_channel.py [hunk 1]
```diff
@@ -197,7 +198,16 @@ def _consume_request_iterator(
...
 
   def consume_request_iterator():
-    for request in request_iterator:
+    while True:
+      try:
+        request = next(request_iterator)
+      except StopIteration:
+        break
+      except Exception as e:
+        logging.exception("Exception iterating requests; {}".format(e))
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_channel.py]: No need to include the exception itself in the message being logged.

File: src/python/grpcio_tests/tests/unit/_invocation_defects_test.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+class FailAfterFewIterationsCounter(object):
+    def __init__(self, high):
+        self._current = 0
+        self._high = high
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self._current > self._high:
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/unit/_invocation_defects_test.py]: Change to `>=` here - it's too weird to pass something a parameter of seven, see it emit eight values, and then fail on the ninth attempt.

File: src/python/grpcio_tests/tests/unit/_invocation_defects_test.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+            raise Exception("This is a deliberate failure in a unit test.")
+        else:
+            self._current += 1
+            return self._current - 1
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/unit/_invocation_defects_test.py]: Doesn't the return value of this method need to be a message of some sort? Not an integer, right? Should `FailAfterFewIterationsCounter` take as a construction parameter the message value to be returned form this method?

A byte string?

**Comment by nathanielmanistaatgoogle** [comment]
Also doesn't the new test need to be added `tests.json`?

**Comment by thunderboltsid** [comment]
@nathanielmanistaatgoogle fixed the issues, take a look now.

File: src/python/grpcio_tests/tests/unit/_invocation_defects_test.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+  def testIteratorStreamRequestStreamResponse(self):
+    requests_iterator = FailAfterFewIterationsCounter(
+      test_constants.STREAM_LENGTH // 2, b'\x07\x08')
+    multi_callable = _stream_stream_multi_callable(self._channel)
+    response_iterator = multi_callable(
+      requests_iterator,
+      metadata=(('test', 'IteratorStreamRequestStreamResponse'),))
+
+    with self.assertRaises(grpc.RpcError):
+      for _ in range(test_constants.STREAM_LENGTH // 2):
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/unit/_invocation_defects_test.py]: Missing ` + 1` here. Just to go over the race condition again:

* Test thread starts RPC, passing a `FailAfterFewIterationsCounter(N)` which when iterated over will emit N messages and then raise an exception when asked for the N+1th message.
* Invocation-side gRPC Python spawns request-consumer thread.
* Request-consumer thread consumes exactly N request values from your `FailAfterFewIterationsCounter`, sending them to the server.
* Request-consumer thread is preempted.
* Server responds to N request values by sending N response values.
* Test thread enters `with` statement, executes entire `for` statement (which loops exactly N times), and exits `with` statement without an exception.
* On exiting `with` statement without an exception, test fails.
* Request consumption thread gets to run again, attempts to get N+1th request, and sees an exception raised by your `FailAfterFewIterationsCounter`.

Reread the steps a few times and tell me where I'm wrong?

File: src/python/grpcio/grpc/_channel.py [hunk 5]
```diff
@@ -197,7 +198,16 @@ def _consume_request_iterator(
...
 
   def consume_request_iterator():
-    for request in request_iterator:
+    while True:
+      try:
+        request = next(request_iterator)
+      except StopIteration:
+        break
+      except Exception as e:
+        logging.exception("Exception iterating requests")
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_channel.py]: Let's keep the exclamation point at the end of this message.

Also: whoa, I'm a hypocrite! ☹️ The other calls to `logging.exception` throughout the codebase do include the exception in the format string. 😦

My apologies for providing poor examples to follow.

File: src/python/grpcio_tests/tests/tests.json [hunk 6]
```diff
@@ -25,7 +25,8 @@
   "_implementations_test.CallCredentialsTest",
   "_implementations_test.ChannelCredentialsTest",
   "_insecure_interop_test.InsecureInteropTest",
-  "_invalid_metadata_test.InvalidMetadataTest"
+  "_invalid_metadata_test.InvalidMetadataTest",
```
Comment (thunderboltsid) [inline src/python/grpcio_tests/tests/tests.json]: @nathanielmanistaatgoogle It's an artefact of the string missing a comma delimiter at the end of the line on the current v1.0.x head.

Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/tests.json]: Wholey Fish. Well, okay.

Comment (soltanmm-google) [inline src/python/grpcio_tests/tests/tests.json]: Er, hm. That makes this file invalid JSON. Might be worth just patching that in another PR as soon as possible.

**Comment by nathanielmanistaatgoogle** [comment]
One thing blocking this is the merge of [pull request 9156](https://github.com/grpc/grpc/pull/9156).

Another might be that Windows `_exit_test` result ([dbg](https://grpc-testing.appspot.com/job/gRPC_pr_win/1501/testReport/junit/(root)/python_windows_dbg/py27_test__exit_test_ExitTest/), [opt](https://grpc-testing.appspot.com/job/gRPC_pr_win/1501/testReport/junit/(root)/python_windows_opt/py27_test__exit_test_ExitTest/)) - @kpayson64, what do you make of it?

**Comment by kpayson64** [comment]
I disabled exit_test a while back because it was flaky.  I'm not sure why its re-appearing here.

**Comment by nathanielmanistaatgoogle** [comment]
@kpayson64: disabled `_exit_test` on both branches or just `master`? This is a `v1.0.x` pull request. I'll kick it for a rerun and see what comes up but do note that the error text introduced in this changelist appears in the log of the failed test.

@thunderboltsid: does [this](https://github.com/grpc/grpc/blob/3fa20cffc6d7f46a6df262a0c8e155619bb7ed39/src/python/grpcio_tests/tests/unit/_exit_scenarios.py#L243) look like a place where we failed in this pull request to update a iterable to an iterator? @kpayson64: why `[REQUEST] * test_constants.STREAM_LENGTH` [there](https://github.com/grpc/grpc/blob/3fa20cffc6d7f46a6df262a0c8e155619bb7ed39/src/python/grpcio_tests/tests/unit/_exit_scenarios.py#L243) rather than `infinite_request_iterator()`?

**Comment by kpayson64** [comment]
@nathanielmanistaatgoogle 
I probably never disabled it no v1.0.x.  I'm not sure why I didn't use ```infinite_request_iterator()``` in that scenario.

On closer inspection, is this a breaking change?  Before we could pass a list as a stream argument, and now we can't.  Even if it doesn't match the type spec exactly, it was all over our test code, and I feel like doing this will pull the rug out from under a number of users.  At a bare minimum, I would _strongly_ recommend not doing this on the 1.0 branch.

**Comment by thunderboltsid** [comment]
@nathanielmanistaatgoogle that seems plausible. I casted iterables to iterators in most places but apparently this one went unnoticed. that's fixed. I partially agree with @kpayson64 that this is a breaking change but the idea is that this PR should hopefully cover all segments of test cases where we were using iterables. We sent around an email on the grpcio mailing list regarding the issue but it went by without any activity.

**Comment by nathanielmanistaatgoogle** [comment]
@thunderboltsid: aww crud this needs conflict resolution and GitHub's online conflict resolution UI is giving me an infinite spinner. I strongly feel that we are [this close](https://www.google.com/search?q=this+close&tbm=isch) to getting this finally integrated.

@kpayson64: this is not a breaking change; passing a iterable rather than an iterator has always been misuse of the API. The earlier we drop _the appearance_ of support, the fewer users out from under whom we pull the rug. I suspect that most mistakes have, like all of ours, been made in tests; I'm not yet aware of any real-world use case in which all request messages of a request-streaming method are known at RPC-invocation-time.

**Comment by kpayson64** [comment]
I can sympathize with the argument that fixing this sooner rather than later is better, and although I disagree with the decision to explicitly forbid lists, I'll defer that decision.

What I can't support is doing this on the 1.0.x branch.  This is not a bug fix, and is far more likely to break someones stable (albeit incorrect) code then to fix an issue.

@nicolasnoble Do you have any thoughts on this?

**Comment by thunderboltsid** [comment]
@nathanielmanistaatgoogle Yeah, I tried it too. It's broken. Anyways, I fixed it now. The old school way still works. :)

**Comment by nathanielmanistaatgoogle** [comment]
@kpayson64: I'm going to merge this but I want you to know that I reflected on your pushback for a few days. My perspective remains that it's better to break behavior that could be misconstrued as supported behavior as early as possible when as few developer-users as possible are mistakenly relying on it. I invite you to feel likewise... but also see the appeal of the other side. Also this is a bug fix for issues [8231](https://github.com/grpc/grpc/issues/8231) and [8454](https://github.com/grpc/grpc/issues/8454).

**Comment by nathanielmanistaatgoogle** [comment]
Known failures:

* [Tsan C++](https://github.com/grpc/grpc/issues/9046)
* [gRPC_performance_pull_requests](https://github.com/grpc/grpc/issues/9047)

**Comment by kpayson64** [comment]
Thanks for the consideration.  I understand that not every decision will be unanimous.
